src/development/cplusplus-codeguide.rst

```diff
@@ -0,0 +1,296 @@
+.. doctest-skip-all
+.. _code-guide:
+
+.. todo::
+    - Pretty much everythin
+    - Actual code guides
+
+************************
+C++ Coding Guidelines
+************************
+This section describes requirements and guidelines for development and testing of a new C++ project on GitLab. The COntinuous Integration tools are GitLab specific - but most of the processes could be easily moved to a Jenkins Pipeline if required. 
+
+.. contents:: Table of Contents
+
+An Example C++ Project
+======================
+
+We have created a `skeleton C++ project
+<https://gitlab.com/steve-ord/cpp_template>`_. Which should provide a full introduction to the various
+recommendations and requirements for the development of C++. The philosophy behind the development of this
+template was to demonstrate one way to meet the project guidelines. There are probably as many ways to organise C++ projects
+as there are developers there are sure to be some contraversial design descisions made. 
+
+This projects demonstrates a recommended
+project layout. It also demonstrates how to implement the following recommended C++ project development features. 
+
+* Continous integration (CI) setup using Gitlab 
+* Cmake as a build tool.
+* GoogleTest framework and example unit testing.
+* C++ linting using clang for stylistic errors.
+* Also test running under valgrind for memory errors.
+* gcov to measure test coverage 
+
+All building and testing is done within a docker container.
+
+Project Layout
+--------------
+Top Level
+^^^^^^^^^
+We have a top level of the project containing:
+
+CMakeLists.txt
+""""""""""""""
+        We have built this template using CMake. The structure of this file
+        will be discussed in `Building the Project`_.   
+
+LICENSE
+"""""""
+        All projects *must* have a license. We have include the recommended BSD
+        3 clause license template *please fill it in*.
+
+README.md
+"""""""""
+        Should at the very least provide a brief discription, installation
+        instructions, pre-requisites
+
+src
+"""
+        This is the top of `the source tree`. It does not have to be called src.
+        But we recommend that the source tree has its head here. The first
+        namespace being the level below this one.
+
+version.txt
+"""""""""""
+        In this template the version file is parsed by the CMakeLists file during the 
+        build process. This specific functionality is not required, but clear versioning is.
+        We recommend - but do not mandate `Semantic Versioning <https://semver.org>`_.
```

Comment:
We mandate semantic versioning. Bullet 4 of https://developer.skatelescope.org/en/latest/development/fundamental_sw_requirements.html#construction